### PR TITLE
[4/4] Intel GPU Runtime Upstreaming for Device

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118613
* #118528
* #118523
* #118091
* #117734
* #117619
* #117611
* __->__ #116869
* #119248

# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), as mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), this last PR  covers the changes under lazy initialization.

# Design
This PR primarily offers the support of multi-processing via lazy initialization. We lazily initialize our runtime avoiding initializing XPU until the first time it is accessed. In our design, we extend `cuda_lazy_init` to `device_lazy_init` which is a device-agnostic API that can support any backend. And change `maybe_initialize_cuda` to `maybe_initialize_device` to support lazy initialization for both CUDA and XPU while maintaining scalability.

# Additional Context
We adopt a similar design to CUDA. So we share some code with CUDA.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: torch/csrc/xpu/Module.cpp [hunk 1]
```diff
@@ -4,13 +4,44 @@
...
 
+static bool in_bad_fork = false; // True for children forked after xpu init
+
+#ifndef WIN32
+// Called in the forked child if xpu has already been initialized
+static void forked_child() {
+  in_bad_fork = true;
+  torch::utils::set_requires_device_init(at::kXPU, true);
+}
+#endif
```
[the remark this answers is not in the document]
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp:25]: done.

File: torch/csrc/xpu/Module.cpp [hunk 2]
```diff
@@ -4,13 +4,44 @@
 #include <c10/xpu/XPUFunctions.h>
 #include <torch/csrc/Module.h>
 #include <torch/csrc/THP.h>
+#include <torch/csrc/utils/device_lazy_init.h>
 #include <torch/csrc/utils/python_numbers.h>
 #include <torch/csrc/utils/python_strings.h>
 
+#ifndef WIN32
+#include <pthread.h>
+#endif
```
[the remark this answers is not in the document]
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp:13]: done.

File: torch/csrc/xpu/Module.cpp [hunk 3]
```diff
@@ -4,13 +4,44 @@
...
+
+// Should be called before the first xpu call. It is mainly called in lazy_init.
+// Note: This is distinct from initExtension because a stub xpu implementation
+// has some working functions (e.g. device_count) but cannot fully initialize.
+static void poison_fork() {
+#ifndef WIN32
+  static c10::once_flag flag;
+  c10::call_once(flag, [] { pthread_atfork(nullptr, nullptr, forked_child); });
+#endif
```
[the remark this answers is not in the document]
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp]: done.

**Review by EikanWang — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Comment by ezyang** [comment]
Since @malfet is dealing with the first PR, I'm letting him deal with the rest of the stack

File: torch/csrc/utils/device_lazy_init.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+static void maybe_initialize_device(at::Device& device) {
+  if (device.is_cuda() || device.is_xpu()) {
+    device_lazy_init(device.type());
+  }
+}
+
+static void maybe_initialize_device(const at::TensorOptions& options) {
+  if (options.device().is_cuda() || options.device().is_xpu()) {
+    torch::utils::device_lazy_init(options.device().type());
```
Comment (gujinghui) [inline torch/csrc/utils/device_lazy_init.h:37]: should align coding style with above func? Both use torch::utils:: or remove it?

File: tools/autograd/templates/python_variable_methods.cpp [hunk 5]
```diff
@@ -975,8 +976,8 @@ static PyObject * THPVariable_to(PyObject* self, PyObject* args, PyObject* kwarg
   auto copy = std::get<3>(parsed);
   auto opt_memory_format = std::get<4>(parsed);
   auto& self_ = THPVariable_Unpack(self);
-  if (device && device->is_cuda()) {
-    torch::utils::cuda_lazy_init();
+  if (device) {
```
Comment (gujinghui) [inline tools/autograd/templates/python_variable_methods.cpp:979]: should we move nullptr check into maybe_initialize_device?
then, the coding style is unified in all scenarios?

File: torch/csrc/utils/device_lazy_init.h (hunk 4 above)
Comment (guangyey) [reply torch/csrc/utils/device_lazy_init.h:37]: remove torch::utils:: to keep the same coding style.

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

File: torch/csrc/utils/device_lazy_init.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * initialized when the first runtime API is requested.
+ *
+ * Here are some common ways that a device object may be retrieved:
+ *   - You call getNonVariableType or getNonVariableTypeOpt
+ *   - You call toBackend() on a Type
+ *
+ * It's important to do this correctly, because if you forget to add it you'll
+ * get an oblique error message seems like "Cannot initialize CUDA without
+ * ATen_cuda library" if you try to use CUDA functionality from a CPU-only
+ * build, which is not good UX.
```
Comment (gujinghui) [inline torch/csrc/utils/device_lazy_init.h:23]: This comments are only for CUDA-bias scenario?

**Review by gujinghui — CHANGES_REQUESTED** [review]
Pls correct the commnets.
https://github.com/pytorch/pytorch/pull/116869/files#r1469040788

File: torch/csrc/utils/device_lazy_init.h (hunk 6 above)
Comment (guangyey) [reply torch/csrc/utils/device_lazy_init.h]: Done.

**Review by gujinghui — APPROVED** [review]

File: test/test_xpu.py [hunk 7]
```diff
@@ -46,6 +53,33 @@ def test_get_device_properties(self):
         self.assertTrue(device_capability["max_work_group_size"] > 0)
         self.assertTrue(device_capability["max_num_sub_groups"] > 0)
 
+    @unittest.skipIf(
+        TEST_WITH_TSAN,
+        "TSAN is not fork-safe since we're forking in a multi-threaded environment",
+    )
+    @unittest.skipIf(IS_WINDOWS, "not applicable to Windows (only fails with fork)")
```
Comment (malfet) [inline test/test_xpu.py:60]: You are not testing it with TSAN nor on Windows, why add those guards?

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for refactoring, but do you mind moving `cuda_lazy_init.h`->`device_lazy_init.h` rename to a separate PR (and I'll quickly approve it)

**Comment by guangyey** [comment]
> Thank you for refactoring, but do you mind moving `cuda_lazy_init.h`->`device_lazy_init.h` rename to a separate PR (and I'll quickly approve it)

OK, I will refactor it soon.

File: test/test_xpu.py (hunk 7 above)
Comment (guangyey) [reply test/test_xpu.py:60]: Yes, you are right. Removed.

**Comment by guangyey** [comment]
@malfet I seperate code change about lazy initialization to another PR, please help review  [refactor lazy init to device-agnostic](https://github.com/pytorch/pytorch/pull/118846)

**Comment by guangyey** [comment]
@malfet , may I know if you have additional comments?

**Comment by EikanWang** [comment]
@malfet , it would be appreciated if you could review this PR again.

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**